ios/browser/api/bookmarks/brave_bookmarks_observer.mm

```diff
@@ -0,0 +1,170 @@
+#include "brave/ios/browser/api/bookmarks/brave_bookmarks_observer.h"
+#include "brave/ios/browser/api/bookmarks/brave_bookmarks_api.h"
+
+#include <memory>
+#include "base/compiler_specific.h"
+#include "base/check.h"
+#include "base/notreached.h"
+#include "components/bookmarks/browser/bookmark_model.h"
+#include "components/bookmarks/browser/bookmark_model_observer.h"
+
+#if !defined(__has_feature) || !__has_feature(objc_arc)
+#error "This file requires ARC support."
+#endif
+
+@interface BookmarkNode(Private)
+- (instancetype)initWithNode:(const bookmarks::BookmarkNode *)node model:(bookmarks::BookmarkModel *)model;
+@end
+
+namespace brave {
+namespace ios {
+class BookmarkModelListener : public bookmarks::BookmarkModelObserver {
+ public:
+  explicit BookmarkModelListener(id<BookmarkModelObserver> observer,
+                               bookmarks::BookmarkModel* model);
+  ~BookmarkModelListener() override;
+
+ private:
+  void BookmarkModelLoaded(bookmarks::BookmarkModel* model, bool ids_reassigned) override;
+  void BookmarkModelBeingDeleted(bookmarks::BookmarkModel* model) override;
+  void BookmarkNodeMoved(bookmarks::BookmarkModel* model,
+                         const bookmarks::BookmarkNode* old_parent,
+                         size_t old_index,
+                         const bookmarks::BookmarkNode* new_parent,
+                         size_t new_index) override;
+  void BookmarkNodeAdded(bookmarks::BookmarkModel* model,
+                         const bookmarks::BookmarkNode* parent,
+                         size_t index) override;
+  void BookmarkNodeRemoved(bookmarks::BookmarkModel* model,
+                           const bookmarks::BookmarkNode* parent,
+                           size_t old_index,
+                           const bookmarks::BookmarkNode* node,
+                           const std::set<GURL>& removed_urls) override;
+  void BookmarkNodeChanged(bookmarks::BookmarkModel* model,
+                           const bookmarks::BookmarkNode* node) override;
+  void BookmarkNodeFaviconChanged(bookmarks::BookmarkModel* model,
+                                  const bookmarks::BookmarkNode* node) override;
+  void BookmarkNodeChildrenReordered(bookmarks::BookmarkModel* model,
+                                     const bookmarks::BookmarkNode* node) override;
+  void BookmarkAllUserNodesRemoved(bookmarks::BookmarkModel* model,
+                                   const std::set<GURL>& removed_urls) override;
+
+  __strong id<BookmarkModelObserver> observer_;
+  bookmarks::BookmarkModel* model_;  // weak
```

Comment:
Is this weakified somewhere else? How does C++/Chromium handle weak pointers

Reply:
this would normally be labelled `Not owned`, but I don't think we want this at all because the BookmarkModel is passed into every method. There's no reason to keep a local reference

Reply:
@kylehickinson just for reference https://chromium.googlesource.com/chromium/src/+/master/base/memory/weak_ptr.h